File: src/Core/DialerProxyBuilder.cs
using System.Text;
using System.Text.Json;
using System.Text.RegularExpressions;

namespace CodexSwitch.Core;

public sealed record DialerProxyInput(
    string RelayYaml,
    string HomeServer,
    string HomePort,
    string HomeUsername,
    string HomePassword,
    string RelayGroup = "relay-group",
    string TargetName = "target-socks5",
    int HttpPort = 1990,
    int SocksPort = 1991,
    string Controller = "127.0.0.1:1993",
    string LimiterListen = "127.0.0.1:1994",
    int MaxConcurrent = 8,
    int MaxInflightDials = 1,
    int DialIntervalMs = 250,
    int QueueWaitS = 8);

public sealed record DialerProxyFiles(string Yaml, string LimiterJson, string LimiterPython);

public static class DialerProxyBuilder
{
    private static readonly Regex RelayName = new(@"^\s*-\s*name:\s*[""']?([^""'\s#]+)", RegexOptions.Multiline);
    private static readonly Regex IPv4 = new(@"^(?:\d{1,3}\.){3}\d{1,3}$");

    public static IReadOnlyList<string> ExtractRelayNames(string relayYaml)
    {
        var names = new List<string>();
        var seen = new HashSet<string>(StringComparer.Ordinal);
        foreach (Match match in RelayName.Matches(relayYaml ?? ""))
        {
            var name = match.Groups[1].Value.Trim();
            if (name.Length == 0 || !seen.Add(name)) continue;
            names.Add(name);
        }
        return names;
    }

    public static string NormalizeRelayYaml(string relayYaml)
    {
        var text = (relayYaml ?? "").Replace("\r\n", "\n").Trim();
        if (text.StartsWith("proxies:", StringComparison.Ordinal))
        {
            var nl = text.IndexOf('\n');
            text = nl < 0 ? "" : text[(nl + 1)..].Trim();
        }
        return text;
    }

    public static DialerProxyFiles Build(DialerProxyInput input)
    {
        var relay = NormalizeRelayYaml(input.RelayYaml);
        var names = ExtractRelayNames(relay);
        if (names.Count == 0) throw new InvalidOperationException("请粘贴至少一个中转节点（需要 `- name:`）。");
        var homeServer = (input.HomeServer ?? "").Trim();
        var homePort = (input.HomePort ?? "").Trim();
        if (homeServer.Length == 0 || homePort.Length == 0)
            throw new InvalidOperationException("请填写美国家宽 SOCKS5 的地址和端口。");

        var listen = SplitHostPort(input.LimiterListen, "127.0.0.1", "1994");
        var via = $"127.0.0.1:{input.SocksPort}";
        var group = string.IsNullOrWhiteSpace(input.RelayGroup) ? "relay-group" : input.RelayGroup.Trim();
        var target = string.IsNullOrWhiteSpace(input.TargetName) ? "target-socks5" : input.TargetName.Trim();
        var loop = IPv4.IsMatch(homeServer)
            ? $"IP-CIDR,{homeServer}/32,{group},no-resolve"
            : $"DOMAIN,{homeServer},{group}";

        var yaml = new StringBuilder();
        yaml.AppendLine($"port: {input.HttpPort}");
        yaml.AppendLine($"socks-port: {input.SocksPort}");
        yaml.AppendLine("allow-lan: true");
        yaml.AppendLine("mode: rule");
        yaml.AppendLine("log-level: info");
        yaml.AppendLine($"external-controller: {YamlScalar(input.Controller)}");
        yaml.AppendLine("ipv6: false");
        yaml.AppendLine("tcp-concurrent: false");
        yaml.AppendLine("keep-alive-idle: 15");
        yaml.AppendLine("keep-alive-interval: 15");
        yaml.AppendLine("proxies:");
        yaml.AppendLine(IndentRelay(relay));
        yaml.AppendLine($"  - name: {YamlScalar(target)}");
        yaml.AppendLine("    type: socks5");
        yaml.AppendLine($"    server: {listen.Host}");
        yaml.AppendLine($"    port: {listen.Port}");
        yaml.AppendLine("    udp: false");
        yaml.AppendLine("    ip-version: ipv4");
        yaml.AppendLine("proxy-groups:");
        yaml.AppendLine($"  - name: {YamlScalar(group)}");
        yaml.AppendLine("    type: select");
        yaml.AppendLine("    proxies:");
        foreach (var name in names) yaml.AppendLine($"      - {YamlScalar(name)}");
        yaml.AppendLine("  - name: Proxy");
        yaml.AppendLine("    type: select");
        yaml.AppendLine("    proxies:");
        foreach (var name in names) yaml.AppendLine($"      - {YamlScalar(name)}");
        yaml.AppendLine($"      - {YamlScalar(target)}");
        yaml.AppendLine("rules:");
        yaml.AppendLine($"  - {loop}");
        foreach (var rule in BypassRelay) yaml.AppendLine($"  - {rule},{group}");
        foreach (var rule in BypassDirect) yaml.AppendLine($"  - {rule},DIRECT");
        foreach (var rule in AiRelay) yaml.AppendLine($"  - {rule},{target}");
        yaml.AppendLine("  - MATCH,DIRECT");

        var limiter = new Dictionary<string, object>
        {
            ["listen"] = $"{listen.Host}:{listen.Port}",
            ["via"] = via,
            ["upstream"] = $"{homeServer}:{homePort}",
            ["username"] = input.HomeUsername ?? "",
            ["password"] = input.HomePassword ?? "",
            ["max_concurrent"] = input.MaxConcurrent,
            ["max_inflight_dials"] = input.MaxInflightDials,
            ["dial_interval_ms"] = input.DialIntervalMs,
            ["queue_wait_s"] = input.QueueWaitS,
            ["handshake_timeout_s"] = 20
        };

        return new DialerProxyFiles(yaml.ToString(), JsonSerializer.Serialize(limiter, Json) + "\n", LimiterPython());
    }

    public static string LimiterPython()
    {
        using var stream = typeof(DialerProxyBuilder).Assembly.GetManifestResourceStream("CodexSwitch.Core.socks-limiter.py")
            ?? throw new InvalidOperationException("缺少内置 socks-limiter.py。");
        using var reader = new StreamReader(stream, Encoding.UTF8);
        return reader.ReadToEnd();
    }

    private static readonly string[] BypassRelay =
    [
        "DOMAIN-SUFFIX,docker.io",
        "DOMAIN,api.github.com",
        "DOMAIN-SUFFIX,github.com",
        "DOMAIN-SUFFIX,githubusercontent.com",
        "DOMAIN-SUFFIX,github.io"
    ];
    private static readonly string[] BypassDirect =
    [
        "DOMAIN-SUFFIX,apple-relay.apple.com",
        "DOMAIN-SUFFIX,apple-relay.fastly-edge.com",
        "DOMAIN-SUFFIX,apple-relay.cloudflare.com",
        "DOMAIN-SUFFIX,guzzoni.apple.com"
    ];
    private static readonly string[] AiRelay =
    [
        "DOMAIN-SUFFIX,openai.com",
        "DOMAIN-SUFFIX,chatgpt.com",
        "DOMAIN-SUFFIX,ai.com",
        "DOMAIN-SUFFIX,anthropic.com",
        "DOMAIN-SUFFIX,claude.ai",
        "DOMAIN-SUFFIX,claude.com",
        "DOMAIN-SUFFIX,x.ai",
        "DOMAIN-SUFFIX,grok.com",
        "DOMAIN-SUFFIX,googleapis.com",
        "DOMAIN-SUFFIX,google.com",
        "DOMAIN-SUFFIX,gemini.google.com",
        "DOMAIN-SUFFIX,cursor.sh",
        "DOMAIN-SUFFIX,githubcopilot.com"
    ];

    private static readonly JsonSerializerOptions Json = new() { WriteIndented = true };

    private static (string Host, string Port) SplitHostPort(string value, string host, string port)
    {
        var text = (value ?? "").Trim();
        if (text.Length == 0) return (host, port);
        var idx = text.LastIndexOf(':');
        if (idx <= 0) return (text, port);
        return (text[..idx], text[(idx + 1)..]);
    }

    private static string IndentRelay(string relay)
    {
        var lines = relay.Split('\n');
        var built = new StringBuilder();
        foreach (var line in lines)
        {
            if (line.Length == 0) { built.AppendLine(); continue; }
            built.AppendLine(line.StartsWith(' ') || line.StartsWith('\t') ? line : "  " + line);
        }
        return built.ToString().TrimEnd();
    }

    private static string YamlScalar(string value)
    {
        if (value.Length == 0) return "\"\"";
        if (Regex.IsMatch(value, @"^[A-Za-z0-9_./-]+$") && !Regex.IsMatch(value, "^(true|false|null)$", RegexOptions.IgnoreCase))
            return value;
        return "\"" + value.Replace("\\", "\\\\").Replace("\"", "\\\"") + "\"";
    }
}

File: src/macOS/Sources/CodexSwitchCore/DialerProxyBuilder.swift
import Foundation

public struct DialerProxyInput: Sendable {
    public var relayYaml: String
    public var homeServer: String
    public var homePort: String
    public var homeUsername: String
    public var homePassword: String
    public var relayGroup: String
    public var targetName: String
    public var httpPort: Int
    public var socksPort: Int
    public var controller: String
    public var limiterListen: String
    public var maxConcurrent: Int
    public var dialIntervalMs: Int
    public var queueWaitS: Int
    public init(relayYaml: String, homeServer: String, homePort: String, homeUsername: String = "", homePassword: String = "",
                relayGroup: String = "relay-group", targetName: String = "target-socks5", httpPort: Int = 1990, socksPort: Int = 1991,
                controller: String = "127.0.0.1:1993", limiterListen: String = "127.0.0.1:1994", maxConcurrent: Int = 8,
                dialIntervalMs: Int = 250, queueWaitS: Int = 8) {
        self.relayYaml = relayYaml; self.homeServer = homeServer; self.homePort = homePort
        self.homeUsername = homeUsername; self.homePassword = homePassword
        self.relayGroup = relayGroup; self.targetName = targetName
        self.httpPort = httpPort; self.socksPort = socksPort; self.controller = controller
        self.limiterListen = limiterListen; self.maxConcurrent = maxConcurrent
        self.dialIntervalMs = dialIntervalMs; self.queueWaitS = queueWaitS
    }
}

public struct DialerProxyFiles: Sendable {
    public let yaml: String
    public let limiterJSON: String
    public let limiterPython: String
}

public enum DialerProxyBuilder {
    public static func extractRelayNames(_ yaml: String) -> [String] {
        var names: [String] = []
        for line in yaml.split(separator: "\n", omittingEmptySubsequences: false) {
            let trimmed = line.trimmingCharacters(in: .whitespaces)
            guard trimmed.hasPrefix("- name:") || trimmed.hasPrefix("-name:") else { continue }
            var value = trimmed.replacingOccurrences(of: "- name:", with: "").replacingOccurrences(of: "-name:", with: "")
                .trimmingCharacters(in: .whitespaces.union(CharacterSet(charactersIn: "\"'")))
            if let hash = value.firstIndex(of: "#") { value = String(value[..<hash]).trimmingCharacters(in: .whitespaces) }
            if !value.isEmpty, !names.contains(value) { names.append(value) }
        }
        return names
    }

    public static func build(_ input: DialerProxyInput) throws -> DialerProxyFiles {
        var relay = input.relayYaml.replacingOccurrences(of: "\r\n", with: "\n").trimmingCharacters(in: .whitespacesAndNewlines)
        if relay.hasPrefix("proxies:"), let nl = relay.firstIndex(of: "\n") {
            relay = String(relay[relay.index(after: nl)...]).trimmingCharacters(in: .whitespacesAndNewlines)
        }
        let names = extractRelayNames(relay)
        guard !names.isEmpty else { throw StoreError("请粘贴至少一个中转节点（需要 `- name:`）。") }
        let homeServer = input.homeServer.trimmingCharacters(in: .whitespaces)
        let homePort = input.homePort.trimmingCharacters(in: .whitespaces)
        guard !homeServer.isEmpty, !homePort.isEmpty else { throw StoreError("请填写美国家宽 SOCKS5 的地址和端口。") }
        let listen = splitHostPort(input.limiterListen, "127.0.0.1", "1994")
        let group = input.relayGroup.isEmpty ? "relay-group" : input.relayGroup
        let target = input.targetName.isEmpty ? "target-socks5" : input.targetName
        let ipv4 = homeServer.range(of: #"^(?:\d{1,3}\.){3}\d{1,3}$"#, options: .regularExpression) != nil
        let loop = ipv4 ? "IP-CIDR,\(homeServer)/32,\(group),no-resolve" : "DOMAIN,\(homeServer),\(group)"
        var yaml = ""
        yaml += "port: \(input.httpPort)\nsocks-port: \(input.socksPort)\nallow-lan: true\nmode: rule\nlog-level: info\n"
        yaml += "external-controller: \(scalar(input.controller))\nipv6: false\ntcp-concurrent: false\nkeep-alive-idle: 15\nkeep-alive-interval: 15\nproxies:\n"
        yaml += indent(relay) + "\n"
        yaml += "  - name: \(scalar(target))\n    type: socks5\n    server: \(listen.host)\n    port: \(listen.port)\n    udp: false\n    ip-version: ipv4\n"
        yaml += "proxy-groups:\n  - name: \(scalar(group))\n    type: select\n    proxies:\n"
        for name in names { yaml += "      - \(scalar(name))\n" }
        yaml += "  - name: Proxy\n    type: select\n    proxies:\n"
        for name in names { yaml += "      - \(scalar(name))\n" }
        yaml += "      - \(scalar(target))\nrules:\n  - \(loop)\n"
        for rule in ["DOMAIN-SUFFIX,docker.io", "DOMAIN,api.github.com"] { yaml += "  - \(rule),\(group)\n" }
        for rule in ["DOMAIN-SUFFIX,openai.com", "DOMAIN-SUFFIX,chatgpt.com", "DOMAIN-SUFFIX,anthropic.com",
                     "DOMAIN-SUFFIX,claude.ai", "DOMAIN-SUFFIX,x.ai", "DOMAIN-SUFFIX,grok.com"] {
            yaml += "  - \(rule),\(target)\n"
        }
        yaml += "  - MATCH,DIRECT\n"
        let limiter: [String: Any] = [
            "listen": "\(listen.host):\(listen.port)", "via": "127.0.0.1:\(input.socksPort)",
            "upstream": "\(homeServer):\(homePort)", "username": input.homeUsername, "password": input.homePassword,
            "max_concurrent": input.maxConcurrent, "max_inflight_dials": 1, "dial_interval_ms": input.dialIntervalMs,
            "queue_wait_s": input.queueWaitS, "handshake_timeout_s": 20
        ]
        let json = try JSONSerialization.data(withJSONObject: limiter, options: [.prettyPrinted, .sortedKeys])
        return DialerProxyFiles(yaml: yaml, limiterJSON: String(data: json, encoding: .utf8)! + "\n", limiterPython: limiterPython())
    }

    public static func limiterPython() -> String {
        guard let url = Bundle.module.url(forResource: "socks-limiter", withExtension: "py"),
              let text = try? String(contentsOf: url, encoding: .utf8) else {
            return ""
        }
        return text
    }

    private static func indent(_ relay: String) -> String {
        relay.split(separator: "\n", omittingEmptySubsequences: false).map { line in
            if line.isEmpty { return "" }
            return line.first == " " || line.first == "\t" ? String(line) : "  " + line
        }.joined(separator: "\n")
    }
    private static func scalar(_ value: String) -> String {
        if value.range(of: #"^[A-Za-z0-9_./-]+$"#, options: .regularExpression) != nil { return value }
        return "\"" + value.replacingOccurrences(of: "\\", with: "\\\\").replacingOccurrences(of: "\"", with: "\\\"") + "\""
    }
    private static func splitHostPort(_ value: String, _ host: String, _ port: String) -> (host: String, port: String) {
        guard let idx = value.lastIndex(of: ":"), idx > value.startIndex else { return (host, port) }
        return (String(value[..<idx]), String(value[value.index(after: idx)...]))
    }
}
